state/HoldingList.go

```diff
@@ -0,0 +1,212 @@
+package state
+
+import (
+	"fmt"
+	"github.com/FactomProject/factomd/common/constants"
+	"github.com/FactomProject/factomd/common/messages"
+
+	"github.com/FactomProject/factomd/common/interfaces"
+)
+
+// toggle to disable old/new for testing
+const useNewHolding = true
+
+// This hold a slice of messages dependent on a hash
+type HoldingList struct {
+	holding    map[[32]byte][]interfaces.IMsg
+	s          *State // for debug logging
+	dependents map[[32]byte]bool
+}
+
+func (l *HoldingList) Init(s *State) {
+	l.holding = make(map[[32]byte][]interfaces.IMsg)
+	l.s = s
+	l.dependents = make(map[[32]byte]bool)
+
+	if !useNewHolding {
+		l.s.LogPrintf("newHolding", "DISABLED")
+	}
+}
+
+func (l *HoldingList) Messages() map[[32]byte][]interfaces.IMsg {
+	return l.holding
+}
+
+func (l *HoldingList) GetSize() int {
+	return len(l.dependents)
+}
+
+func (l *HoldingList) Exists(h [32]byte) bool {
+	return l.dependents[h]
+}
+
+// Add a message to a dependent holding list
+func (l *HoldingList) Add(h [32]byte, msg interfaces.IMsg) bool {
+
+	if l.dependents[msg.GetMsgHash().Fixed()] {
+		return false
+	}
+
+	if l.holding[h] == nil {
+		l.holding[h] = []interfaces.IMsg{msg}
+	} else {
+		l.holding[h] = append(l.holding[h], msg)
+	}
+
+	l.dependents[msg.GetMsgHash().Fixed()] = true
+	return true
+}
+
+// get and remove the list of dependent message for a hash
+func (l *HoldingList) Get(h [32]byte) []interfaces.IMsg {
+	rval := l.holding[h]
+	delete(l.holding, h)
+
+	for _, msg := range rval {
+		l.s.LogMessage("newHolding", "DequeueFromDependantHolding()", msg)
+		delete(l.dependents, msg.GetMsgHash().Fixed())
+	}
+	return rval
+}
+
+func (l *HoldingList) ExecuteForNewHeight(ht uint32) {
+	l.s.ExecuteFromHolding(HeightToHash(ht))
+}
+
+// clean stale messages from holding
+func (l *HoldingList) Review() {
+
+	if !useNewHolding && l.GetSize() > 0 {
+		panic("found messages in new-holding while disabled")
+	}
+
+	for h := range l.holding {
+		dh := l.holding[h]
+		if nil == dh {
+			continue
+		}
+		for _, msg := range dh {
+			if l.isMsgStale(msg) {
+				l.Get(h) // remove all from holding
+				l.s.LogMessage("newHolding", "RemoveFromDependantHolding()", msg)
+				continue
+			}
+		}
+	}
+}
+
+func (l *HoldingList) isMsgStale(msg interfaces.IMsg) (res bool) {
+
+	/*
+		REVIEW:
+		Maybe we should treat the message stream as a votes on the "highest known block" where known servers trump unknown servers who disagree?
+
+		Consider setting HKB and HAB when we complete minute 1 of a block to the current leader height.
+		That at least would make us recover from a spoofed ack attack.
+	*/
+
+	switch msg.Type() {
```

Comment:
How do Entry commits/reveals get expired?

Reply:
By timestamp below. Got it :+1:

Reply:
they expire by timeout via timestamp rather than height